### ISSUE TITLE
Track all deployments

### ISSUE THREAD

**Issue by PatrickAlphaC** [ApeWorX/ape#809]
### Elevator pitch:
Right now, we have https://github.com/ApeWorX/ape/issues/530 which is a design to track deployments as such:

```python
from ape import project 

my_contract = account.deploy(project.MyContract)
project.track_deployment(my_contract) 
my_contract = project.deployments(project.MyContract)[-1]
```

The line `project.track_deployment(my_contract)` would add information about this contract deployment to `.build/deployments/<ecosystem>/<network>/track_deployments.json`

This file would look as such:

```json
{
  "MyContract": [
    "0x40Ee4CfF4d103509cEe0720362b8055Ad65f259b",
    .
    .
    .
  ],
  "NextContract:[
  .
  .
  ],
...
}
```

This is how the project would keep track of deployments. 

### Value:

This is for all developers who don't want to have to write custom scripts to keep track of their deployments. 

### Dependencies:

I don't think this has any dependencies.

### Design approach:
The main design is in the elevator pitch. But to continue, we may have to adjust the api for local networks. We'd wouldn't always want the saved deployment to be saved forever on local networks that get destroyed so often:

```python
def track_deployment(contract, persist: bool=true):
```
Or, we could pull from the ape-config.yaml

```yaml
name: apeworx-project
contracts_folder: contracts
networks:
  hardhat:
    persist: false
```

### Task list:

* [ ] Deployments can be tracked
* [ ] Tests for deployments written
* [ ] Documentation updates

### Estimated completion date:
2 weeks

### Design review:

Do not signoff unless:
- 1) agreed the tasks and design approach will achieve acceptance, and
- 2) the work can be completed by one person within the SLA.
Design reviewers should consider simpler approaches to achieve goals.

(Please leave a comment to sign off)

Let me know if this is something I can take a crack at and if the design looks good.

**Comment by fubuloubu** [comment]
The idea behind #530 is to track things that should be stored in a project's permanent package manifest when the project is ready to be published. Along the way, you might make deployments to public networks (like testnets) as well as locally and want to obtain deployments you have made that are currently live (as a side note: in live networks, they only persist for as long as a session persists)

We currently already have a contract cache that stores what contract types have been deployed to what addresses, and it manages all of the complexities of local session-only deployments vs. permanent public network ones:
https://github.com/ApeWorX/ape/blob/9ee05441bff75d37897f24b72016810937dda6f1/src/ape/managers/chain.py#L362

What we don't have is the inverse lookup: based on deployments of a particular contract type that we have made, how do we access a list of those deployments through good UX.

I think this ticket should be redesigned a bit to talk more about accessing that list inside of the `ContractCache` object (e.g. `chain.contracts.<ContractTypeName>[<index_of_deployment>]`) so that the UX for this feature is pretty clear.

**Comment by PatrickAlphaC** [comment]
So, how’s this for an API:

```python

from ape import project, chain

my_contract = account.deploy(project.MyContract)
project.track_deployment(my_contract) 
my_contract = chain.contracts.MyContract[-1]

```

**Comment by fubuloubu** [comment]
> So, how’s this for an API:
> 
> ```python
> from ape import project, chain
> 
> my_contract = account.deploy(project.MyContract)
> project.track_deployment(my_contract) 
> my_contract = chain.contracts.MyContract[-1]
> ```
> 
> or better yet, we just remove the `track_deployment` line totally, and pull from the ape config on whether or not to track (testnets and public networks default to true, local networks default to false)

Yes, we don't need `track_deployment` to use this feature (the use case there is for keeping track of _final_ copies of contracts to be deployed e.g. official releases), maybe let's do this instead:

```py
from ape import project, chain

my_contract = account.deploy(project.MyContract)
my_contract = chain.contracts.find_deployments(project.MyContract)[-1]
# NOTE: Has to do a search over the contract cache
#       to find deployments with matching contract type
```

**Comment by PatrickAlphaC** [comment]
Took a day to play with this... not sure if you'd be interested in me making a draft PR. I'm having a hard time running tests on a sample project. I can't seem to create a project where the `contracts_folder` is in the `tests/functional/data/projects/` area. 

I was thinking I'd do something like:

```python
@pytest.fixture(scope="session")
def project_with_contract():
    yield ProjectManager(APE_PROJECT_FOLDER)
```

So I can run a test like:

```python
def test_find_deployments(chain, project_with_contract):
    # Arrange / Act
    project_with_contract.ApeContract.deploy()
    my_contracts_list = chain.contracts.find_deployments(project.ApeContract)

    # Assert
    assert len(my_contracts_list) == 1
```


And I made a `contracts_map` var that I've written to disc, and access it with:
```python
    def find_deployments(contract_type: ContractType) -> List[ContractInstance]:
        contracts_map = self._load_contracts_mapping()
        network_name = self._network.name.replace("-fork", "")
        ecosystem_name = self._network.ecosystem.name

        return (
            contracts_map.get(ecosystem_name, {}).get(network_name, {}).get(ContractType.__name__)
            or []
        )
```

But I can't get the project manager to recognize that it should point to my `ApeContract` folder I created. It looks like I need to do something with the `ConfigManager`

# TODO

Let me know if you'd like to see a draft PR. If you're too busy, all good. Not sure how close or far I am here.

**Comment by PatrickAlphaC** [comment]
Draft PR: https://github.com/ApeWorX/ape/pull/856

Really ugly :/

**Comment by unparalleled-js** [comment]
Renamed the issue because there is a separate ticket for tracking deployments for publishing reasons that goes in the .build folder.

This ticket ended up by closed by #856 and was just for _all_ deployments.

I know the history in confusing but otherwise this is duplicate of https://github.com/ApeWorX/ape/issues/530
